src/data/polynomial/reverse.lean

```diff
@@ -0,0 +1,214 @@
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+
+import data.polynomial.degree.basic
+import data.polynomial.erase_lead
+import data.polynomial.degree.trailing_degree
+
+/-!
+# Reverse of a univariate polynomial
+
+The main definition is `reverse`.  Applying `reverse` to a polynomial `f(x)` produces the
+polynomial `x^f.nat_degree * f(1/x)`.
+
+The main result is that `reverse (f * g) = reverse (f) * reverse (g)`, provided the leading
+coefficients of `f` and `g` do not multiply to zero.
+-/
+
+namespace polynomial
+
+open polynomial finsupp finset --erase_lead
+
+
+variables {R : Type*} [semiring R] {f : polynomial R}
+
+
+namespace rev
+/-- rev_at is a function of two natural variables (N,i).  If i ≤ N, then rev_at N i returns N-i,
+otherwise it returns N.  Essentially, this function is only used for i ≤ N. -/
+def rev_at (N : ℕ) : ℕ → ℕ := λ i : ℕ, ite (i ≤ N) (N-i) i
+
+@[simp] lemma rev_at_invol {N n : ℕ} : rev_at N (rev_at N n) = n :=
+begin
+  unfold rev_at,
+  split_ifs with h j,
+  { exact nat.sub_sub_self h, },
+  { exfalso,
+    apply j,
+    exact nat.sub_le N n, },
+  { refl, },
+end
+
+@[simp] lemma rev_at_small {N n : ℕ} (H : n ≤ N) : rev_at N n = N-n :=
+begin
+  unfold rev_at,
+  split_ifs,
+  refl,
+end
+
+/-- The function `reflect` of a natural number `N` and a polynomial `f`, applies the function
+`rev_at` to the exponents of the terms appearing in the expansion of `f`.  In practice, `reflect`
+is only used when `N` is at least as large as the degree of `f`.  Eventually, it will be used with
+`N` exactly equal to the degree of `f`.  -/
+def reflect : ℕ → polynomial R → polynomial R :=
+ λ N : ℕ, λ f : polynomial R, ⟨ image (rev_at N)  (f.support), λ i : ℕ, f.coeff (rev_at N i),
+ begin
+  intro,
+  rw mem_image,
+  split,
+  { intro h,
+    rcases h with ⟨ a, ha, rfl⟩,
+    rwa [rev_at_invol, ← mem_support_iff_coeff_ne_zero], },
+  { intro,
+    use (rev_at N a),
+    rwa [rev_at_invol, eq_self_iff_true, and_true, mem_support_iff_coeff_ne_zero], },
+end ⟩
+
+@[simp] lemma reflect_zero {n : ℕ} : reflect n (0 : polynomial R) = 0 := rfl
+
+@[simp] lemma reflect_zero_iff {N : ℕ} (f : polynomial R) :
+ reflect N (f : polynomial R) = 0 ↔ f=0 :=
+begin
+  split,
+  { intros a,
+    injection a with f0 f1,
+    rwa [image_eq_empty, support_eq_empty] at f0, },
+  { intro a,
+    rw [a, reflect_zero], },
+end
+
+@[simp] lemma reflect_add (f g : polynomial R) (n : ℕ) :
+ reflect n (f+g) = reflect n f + reflect n g :=
+begin
+  ext1,
+  refl,
+end
+
+@[simp] lemma reflect_smul (f : polynomial R) (r : R) (N : ℕ) :
+ reflect N (C r * f) = C r * (reflect N f) :=
+begin
+  ext1,
+  unfold reflect,
+  rw [coeff_mk, coeff_C_mul, coeff_C_mul, coeff_mk],
+end
+
+@[simp] lemma reflect_C_mul_X_pow (N n : ℕ) {c : R} :
+ reflect N (C c * X ^ n) = C c * X ^ (rev_at N n) :=
+begin
+  ext1 m,
+  unfold reflect,
+  rw coeff_mk,
+  by_cases h : rev_at N n = m,
```

Comment:
I think that I followed your suggestion and obtained this proof

```
  ext,
  rw [reflect_C_mul, coeff_C_mul, coeff_C_mul, coeff_X_pow],
  unfold reflect,
  rw [coeff_mk],
  split_ifs,
  { rw [h, rev_at_invol, coeff_X_pow_self], },
  { rw [not_mem_support_iff_coeff_zero.mp],
    intro a,
    rw [← one_mul (X ^ n), ← C_1] at a,
    apply h,
    rw [← (mem_support_C_mul_X_pow a), rev_at_invol], },
```

I hope that this is what you had in mind!

Reply:
That looks good, thanks! Maybe you can turn the `unfold reflect` into `rw reflect` as well, to save two more lines.

Reply:
Ah, I did not know that you could use ```rw``` instead of ```unfold```!  I used this trick a few more times to absorb ```unfolds``` into longer ```rw```s!